tsdb/head_read.go

```diff
@@ -617,7 +630,7 @@ func (c *safeChunk) Iterator(reuseIter chunkenc.Iterator) chunkenc.Iterator {
 // iterator returns a chunk iterator for the requested chunkID, or a NopIterator if the requested ID is out of range.
 // It is unsafe to call this concurrently with s.append(...) without holding the series lock.
 func (s *memSeries) iterator(id chunks.HeadChunkID, isoState *isolationState, chunkDiskMapper *chunks.ChunkDiskMapper, memChunkPool *sync.Pool, it chunkenc.Iterator) chunkenc.Iterator {
-	c, garbageCollect, err := s.chunk(id, chunkDiskMapper, memChunkPool)
+	c, garbageCollect, err := s.chunk(id, chunkDiskMapper, memChunkPool, nil)
```

Comment:
Why passing `nil` here? I understand `iterator()` itself is documented as unsafe to get called concurrent to `s.append()` but the returned chunk contains the `[]byte` which will be later manipulated by an `s.append()` (after the iterator lock is released (see `safeChunk.Iterator()`). Isn't unsafe what we're doing here?

Reply:
It is safe. At the end we are taking the Iterator() on the chunk. It uses the [newBReader](https://github.com/prometheus/prometheus/blob/86d3d5fec6f9063fb4d26eff931b712a31d7e782/tsdb/chunkenc/bstream.go#L125-L130) underneath which reads the chunk in a safe manner. The returned iterator can only iterate samples and can't ask for []byte. Hence we skip copying the chunk (by passing `nil`) since it is not required.

Reply:
Okay I see what you mean now. Chunk() method returns a safeChunk. And the safeChunk's iterator call is not making use of the chunk that is embedded, instead getting another chunk. I will look into that.